[Workflow] Add generic template for event classes

fractalzombie opened the pull request:
Added generic template for `Event` in Workflow Component

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| License       | MIT

Added generic template for `Event`, `Event::$subject` and `Event::getSubject()` for PHPStorm support.

fractalzombie committed:
[Workflow] Added generic template for `Event` and his inheritors and for `Event::$subject` and `Event::getSubject()`

derrabus commented:
Can you elaborate your motivation for this change? Why do you need these annotations?

derrabus commented:
Also, please target the 6.1 branch.

derrabus commented on src/Symfony/Component/Workflow/Event/Event.php:
```diff
@@ -17,13 +17,18 @@
 class Event extends BaseEvent
 {
     protected $context;
+    /**
+     * @var TSubject
+     */
```
We should rather annotate the constructor parameter.

fractalzombie replied:
Agree...

fractalzombie commented:
> Can you elaborate your motivation for this change? Why do you need these annotations?

For PHPStorm generic support.

derrabus commented on src/Symfony/Component/Workflow/Event/Event.php:
```diff
@@ -17,6 +17,8 @@
 use Symfony\Contracts\EventDispatcher\Event as BaseEvent;
 
 /**
+ * @template TSubject
```
```suggestion
 * @template TSubject of object
```

derrabus commented:
The Psalm error looks related.

derrabus commented:
> > Can you elaborate your motivation for this change? Why do you need these annotations?
> 
> For PHPStorm generic support.

That's a very generic answer (no pun intended). Can you be more specific? How do the annotations help you? What would PhpStorm do differently if we added them to the codebase?

fractalzombie commented:
> > > Can you elaborate your motivation for this change? Why do you need these annotations?
> > 
> > 
> > For PHPStorm generic support.
> 
> That's a very generic answer (no pun intended). Can you be more specific? How do the annotations help you? What would PhpStorm do differently if we added them to the codebase?

What are you mean, when you say "What would PhpStorm do differently if we added them to the codebase?". I think it's not about what PHPStorm do differently, it's about, when we can Annotate Event with `Event<EntityOrModel>` and then we can have intellisense with Storm.

derrabus commented:
> when we can Annotate Event with `Event<EntityOrModel>` and then we can have intellisense with Storm.

That kind-of answers my question, I guess.

Would it make sense to add the template type for the subject to the workflow object as well?

derrabus commented on src/Symfony/Component/Workflow/Event/GuardEvent.php:
```diff
@@ -16,8 +16,10 @@
 use Symfony\Component\Workflow\TransitionBlocker;
 use Symfony\Component\Workflow\TransitionBlockerList;
 use Symfony\Component\Workflow\WorkflowInterface;
-
```
fabbot will keep complaining if you don't add back that blank line. 🙂

fractalzombie replied:
Yep, we can, why not? I will add it...

fractalzombie commented:
> > when we can Annotate Event with `Event<EntityOrModel>` and then we can have intellisense with Storm.
> 
> That kind-of answers my question, I guess.
> 
> Would it make sense to add the template type for the subject to the workflow object as well?

We can add, but in workflow it will be do nothing because it's only be available in library.

derrabus commented on src/Symfony/Component/Workflow/Event/Event.php:
```diff
@@ -43,6 +48,9 @@ public function getMarking()
     }
 
+    /**
+     * @return TSubject|object
```
This does not look correct. We should always get a `TSubject` here. Wasn't that the whole point of the PR?

fractalzombie replied:
Nope, in this case psalm have issue, says that it can be object or Tsubject

ro0NL replied:
What issue? https://psalm.dev/r/4582b1218d

if anything, adding `|object` creates an issue.

fractalzombie replied:
I will remove it now, and we see

fractalzombie replied:
@ro0NL And, here we have the issue https://github.com/symfony/symfony/runs/453131776

ro0NL replied:
Weird. Maybe try to reproduce on psalm.dev, but here psalm seems wrong AFAIK.

stof commented:
To make psalm happy, I think you need to add `@var TSubject` in the phpdoc of the property, so that psalm considers the property as being of type `TSubject` rather than `object`

stof commented on src/Symfony/Component/Workflow/Event/Event.php:
```diff
@@ -17,6 +17,8 @@
 use Symfony\Contracts\EventDispatcher\Event as BaseEvent;
 
 /**
+ * @template TSubject of object
```
Given that the subject is only used in return types and not in argument types (not counting the constructor), this could use `@template-covariant` rather than `@template`, which is more flexible (same for child classes)

fractalzombie replied:
covariant gives error, i replace it with `@template`, for now it's ok and as i see no errors anymore

fractalzombie commented:
> To make psalm happy, I think you need to add `@var TSubject` in the phpdoc of the property, so that psalm considers the property as being of type `TSubject` rather than `object`

You can check upper, I already do this, but @derrabus don't happy with it.

stof commented:
Well, we need to annotate *both* the constructor and the property, because the property inference of psalm is done based on the native argument type rather than the phpdoc type (and in that case, we need the more precise type `TSubject`).

fractalzombie commented:
> Well, we need to annotate _both_ the constructor and the property, because the property inference of psalm is done based on the native argument type rather than the phpdoc type (and in that case, we need the more precise type `TSubject`).

Added, fixed.

fractalzombie commented:
> Well, we need to annotate _both_ the constructor and the property, because the property inference of psalm is done based on the native argument type rather than the phpdoc type (and in that case, we need the more precise type `TSubject`).

I think the last commit is worked

fractalzombie commented:
@derrabus something strange is going on with my fork, i need to recreate, and we can close this and accept this https://github.com/symfony/symfony/pull/44702.
